tough/src/transport.rs

```diff
@@ -1,36 +1,172 @@
-use std::io::Read;
+#[cfg(feature = "http")]
+use crate::{ClientSettings, HttpTransport};
+use dyn_clone::DynClone;
+use snafu::Snafu;
+use std::fmt::Debug;
+use std::io::{ErrorKind, Read};
 use url::Url;
 
 /// A trait to abstract over the method/protocol by which files are obtained.
-pub trait Transport {
-    /// The type of `Read` object that the `fetch` function will return.
-    type Stream: Read;
+///
+/// The trait hides the underlying types involved by returning the `Read` object as a
+/// `Box<dyn Read + Send>` and by requiring concrete type [`TransportError`] as the error type.
+///
+pub trait Transport: Debug + DynClone {
+    /// Opens a `Read` object for the file specified by `url`.
+    fn fetch(&self, url: Url) -> Result<Box<dyn Read + Send>, TransportError>;
+}
 
-    /// The type of error that the `fetch` function will return.
-    type Error: std::error::Error + Send + Sync + 'static;
+// Implement `Clone` for `Transport` trait objects.
+dyn_clone::clone_trait_object!(Transport);
 
-    /// Opens a `Read` object for the file specified by `url`.
-    fn fetch(&self, url: Url) -> Result<Self::Stream, Self::Error>;
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// The kind of error that the transport object experienced during `fetch`.
+///
+/// # Why
+///
+/// Some TUF operations need to know if a [`Transport`] failure is a result of a file not being
+/// found. In particular:
+/// > 5.1.2. Try downloading version N+1 of the root metadata file `[...]` If this file is not
+/// > available `[...]` then go to step 5.1.9.
+///
+/// To distinguish this case from other [`Transport`] failures, we use
+/// `TransportErrorKind::FileNotFound`.
+///
+#[derive(Debug, Copy, Clone)]
+#[non_exhaustive]
+pub enum TransportErrorKind {
+    /// The trait does not handle the URL scheme named in `String`. e.g. `file://` or `http://`.
+    UnsupportedUrlScheme,
+    /// The file cannot be found.
+    FileNotFound,
```

Comment:
```suggestion
    /// The file is not available.
    FileNotAvailable,
```

I feel like the current language implies the HTTP transport should only return this error kind on a 404, which is not what we'll be doing, because the production Bottlerocket repositories (currently) return a 403 for https://updates.bottlerocket.aws/2020-07-07/aws-k8s-1.17/x86_64/3.root.json, because that's what S3 does when the principal making the request does not have `s3:ListObjects` access.

The [TUF specification (v1.0.16)](https://github.com/theupdateframework/specification/blob/fd85a8a12403435d84e202ea8790eb4abad4559d/tuf-spec.md#5-detailed-workflows) is a bit vague here, simply saying "If this file is not available ... then go to step 5.2.9." Replicating the language here makes it clearer what an implementer should do.

As we implement a standard HTTP transport for tough, we have to decide which types of HTTP error codes count as "definitely not available" (e.g. 404, 403, maybe 410 Gone?) vs "could be available, but something is wrong" (e.g. 401 Unauthorized, 500 Internal Server Error), and ideally document that.